### DISCUSSION TITLE
Help! `return` action behaving strange

### DISCUSSION THREAD

**Discussion by RaviRahar** [ibhagwan/fzf-lua#1329] [Q&A] [answered]
I have [`return`](https://github.com/RaviRahar/fzf-combi-mode/blob/26908068545eec03db8e80e2b2269e42bbe23e86/lua/fzf-combi-mode.lua#L500) action in my custom browser_mode. It only works 50% of the time. It stopped working recently only after I updated FzfLua. Something has changed. I looked at commits and can only see one [commit](https://github.com/ibhagwan/fzf-lua/commit/0e08ceab0e354ec9450d93c39aef512a34b17d1b) relating to `default` and `enter` actions. I do not have much time to investigate, can someone familiar with FzfLua's code help me find the problem. I will list "normal behaviour" and "bugged behaviour" below:

- Normal behaviour on pressing 'return', if selection is:
  - file: then open it using `fzf_lua.actions.file_edit`
  - directory: then open it again in browser_mode listing all files and dirs
  
- Bugged behaviour on pressing 'return':
  - if selection is file then it does nothing (probably it is crashing)-
  - Next time when you open browser_mode then `return` starts doing action set for `default`,  i.e., it lists files in parent directory.
  
Video of bug showing both behaviours:

https://github.com/user-attachments/assets/4c949c3b-9d23-4455-94c2-017cf1ac6622

**Note: same config produces both behaviours and it flips randomly. Sometimes it works, other times it doesn't.**
**Also I was lucky to get both behaviours in consecutive runs, it was just a coincidence.**

**Comment by ibhagwan** [comment]
Why do you have both `default` and `return` defined? Isn’t one of these overriding the other?

Can you change `return` with `enter`?

**Comment by RaviRahar** [comment]
> Why do you have both default and return defined? Isn’t one of these overriding the other?

I have set [`fzf = { ["backward-eof"] = "accept" }`](https://github.com/RaviRahar/fzf-combi-mode/blob/26908068545eec03db8e80e2b2269e42bbe23e86/lua/fzf-combi-mode.lua#L619). So that I can program backspace to go to parent directory when query is empty. This way was one of your [suggestions](https://github.com/ibhagwan/fzf-lua/issues/805#issuecomment-1618701571) when I was trying to figure out a way to do it.

> Can you change return with enter?

I have tried it earlier. It didn't work. Now behaviour for `return` is fixed but backspace starts behaving as `return` too. It just performs action specified in `enter`. Because of `backward-eof` set to `accept` and `enter` being used for `default` action (as per your [commit](https://github.com/ibhagwan/fzf-lua/commit/0e08ceab0e354ec9450d93c39aef512a34b17d1b)).

**Comment by ibhagwan** [comment, marked as answer]
> I have set [fzf = { ["backward-eof"] = "accept" }](https://github.com/RaviRahar/fzf-combi-mode/blob/26908068545eec03db8e80e2b2269e42bbe23e86/lua/fzf-combi-mode.lua#L619). So that I can program backspace to go to parent directory when query is empty. This way was one of your https://github.com/ibhagwan/fzf-lua/issues/805#issuecomment-1618701571 when I was trying to figure out a way to do it.

With fzf 0.53 a new `print` action was added that enabled combination of both fzf and fzf-lua's actions so now we could much more easily achieve `select-all+accept` for any action not just the default (and many other more involved setups).

I never realized it `backward-eof` would be used in this way and that there would be a differentiation between the default action and `enter` (or `return`, same thing).

I do have a solution for you though, with https://github.com/ibhagwan/fzf-lua/commit/3d214f8db4a4119723ab5148b7a333aab3fa9063 I made it so all actions that start with `_underscore` won't be defined as an fzf bind so you can bind `backward-eof` to any custom action, for example:
```lua
require("fzf-lua").fzf_exec({ 1, 2, 3, 4 }, {
  keymap = { fzf = { ["backward-eof"] = "print(_myaction)+accept" } },
  actions = {
    ["enter"] = function(sel, _opts)
      print("enter, num sel:", #sel)
    end,
    ["_myaction"] = function(sel, _opts)
      print("_myaction, num sel", #sel)
    end
  }
})
```

**Reply by RaviRahar** [reply]
Thank you for quick fix. It works now after I changed every `default` action used for backspace to custom action with this [commit](https://github.com/RaviRahar/fzf-combi-mode/commit/557a04c6e8d3fe9cbdd72f873c6733f9715fe92c). I have long realized I am missing some basics of fzf. I need to take a week off and read fzf's man. It's just that right now I don't have a week and I keep delaying it.

**Reply by ibhagwan** [reply]
Looks great @RaviRahar, just remember this will only work with fzf >= 0.53.
